Happy Card Writeup (#13)
kitctf/www: _posts/2023-01-08-rwctf2023-happycard.md

- Read eeprom data using (now no longer) out-of-bounds read on `byte[]`
- Read flag 🍉

## Challenge Description

Suppose we have a card and can load something on it.
`nc 198.11.179.216 7788`

When you connect to the service you are asked for your team-id then you're prompted to upload a tar file containing your `.cap` file[s].

The attachment contains
```
  creating: attachments/
  inflating: attachments/Dockerfile
   creating: attachments/files/
  inflating: attachments/files/entrypoint.sh
  inflating: attachments/files/hello.cap
  inflating: attachments/files/java_card_simulator-3_1_0-u5-win-bin-do-b_70-09_mar_2021.msi
  inflating: attachments/files/java_card_tools-win-bin-b_17-06_jul_2021.zip
  inflating: attachments/start.sh
```

So a pretty basic docker setup as well as a `hello.cap`, and a few tools related to `java-card` including the simulator in version 3.1.0u5.

## The entrypoint
```bash
[...]

verifycap() {
	java -Djc.home="$JC_HOME_TOOLS" -classpath "$TOOLS_CP" com.sun.javacard.offcardverifier.Verifier -nobanner $@
}

scriptgen() {
	java -Djc.home="$JC_HOME_SIM" -classpath "$SIM_CP" com.sun.javacard.scriptgen.Main -nobanner $@
}

script=/tmp/script

> verifycap -outfile /tmp/hello.cap.digest /files/hello.cap

Reviewer: ```suggestion
verifycap -outfile /tmp/hello.cap.digest /files/hello.cap # 3
```

> scriptgen -hashfile /tmp/hello.cap.digest -o /tmp/hello.cap.script /files/hello.cap

Reviewer: ```suggestion
scriptgen -hashfile /tmp/hello.cap.digest -o /tmp/hello.cap.script /files/hello.cap # 3
```

cat << EOF > $script
powerup;
output off;
0x00 0xA4 0x04 0x00 0x09 0xA0 0x00 0x00 0x00 0x62 0x03 0x01 0x08 0x01 0x7F;
EOF
FLAG=`echo -n "$FLAG"|perl -lne 'print map {"0x".(unpack "H*",$_)." "} split //, $_;'`

> cat /tmp/hello.cap.script >> $script

Reviewer: ```suggestion
cat /tmp/hello.cap.script >> $script # 3
```

cat << EOF >> $script
0x80 0xB8 0x00 0x00 0x08 0x06 0xAA 0xBB 0xCC 0xDD 0xEE 0xAA 0x00 0x7F;

> 0x00 0xA4 0x04 0x00 0x06 0xAA 0xBB 0xCC 0xDD 0xEE 0xAA 0x7F;

Reviewer: ```suggestion
0x00 0xA4 0x04 0x00 0x06 0xAA 0xBB 0xCC 0xDD 0xEE 0xAA 0x7F; # 4
```
Is this the correct location?

0x88 0x88 0x00 0x00 0x30 $FLAG 0x7f;

> 0x00 0xA4 0x04 0x00 0x09 0xA0 0x00 0x00 0x00 0x62 0x03 0x01 0x08 0x01 0x7F;

Reviewer: ```suggestion
0x00 0xA4 0x04 0x00 0x09 0xA0 0x00 0x00 0x00 0x62 0x03 0x01 0x08 0x01 0x7F; # 6
```
Is this the correct location?

EOF

TMPDIR=/jctmp
mkdir $TMPDIR
cd /upload

> for capfile in *.cap; do

Reviewer: ```suggestion
for capfile in *.cap; do # 7
```

---
layout: post
title: "Real World CTF 2023: Happy-Card Writeup"
categories: writeups
author: I Al Istannen, rad4day
---

[Real World CTF 2023](https://realworldctf.com/challenge) was a jeopardy-style capture-the-flag event.
We participated as part of the [Sauercloud](https://ctftime.org/team/54748) CTF-team.

> ## How solve? Where exploit?

Reviewer: ```suggestion
## How Solve? Where Exploit?
```

- Google for java card stuff
- Find [some usable tooling to build your own applets](https://github.com/martinpaljak/ant-javacard)

> - Enable eeprom dumping in cref (`-o /upload/eeprom`)

Reviewer: ```suggestion
- Enable EEPROM dumping in cref (`-o /upload/eeprom`)
```

- Stumble upon [Sergei Volokitin's Master's Thesis](https://www.ru.nl/publish/pages/769526/sergei_volokitin.pdf) for an overview of vulnerabilities
    - Try multiple, but be caught by byte code verification / type checking
- Stumble upon [PhiAttack](https://cardis2021.its.uni-luebeck.de/papers/CARDIS2021_Dubreuil.pdf)
    - Needs older sdk
    - Implement simple type confusion between `Object` and `byte[]`
- Be confused why some casts work and others dont
    - reverse cref `__saload` and `__baload` read checks
- Use type confusion to cast an `Object` to a `byte[]` array with arbitrary length

> - Read eeprom data using (now no longer) out-of-bounds read on `byte[]`

Reviewer: ```suggestion
- Read EEPROM data using (now no longer) out-of-bounds read on `byte[]`
```

- Read flag 🍉

## Challenge Description

Suppose we have a card and can load something on it.
`nc 198.11.179.216 7788`

When you connect to the service you are asked for your team-id then you're prompted to upload a tar file containing your `.cap` file[s].

The attachment contains
```
  creating: attachments/
  inflating: attachments/Dockerfile
   creating: attachments/files/
  inflating: attachments/files/entrypoint.sh
  inflating: attachments/files/hello.cap
  inflating: attachments/files/java_card_simulator-3_1_0-u5-win-bin-do-b_70-09_mar_2021.msi
  inflating: attachments/files/java_card_tools-win-bin-b_17-06_jul_2021.zip
  inflating: attachments/start.sh
```

So a pretty basic docker setup as well as a `hello.cap`, and a few tools related to `java-card` including the simulator in version 3.1.0u5.

> ## The entrypoint

Reviewer: ```suggestion
## The Entrypoint
```

```bash
[...]

verifycap() {
	java -Djc.home="$JC_HOME_TOOLS" -classpath "$TOOLS_CP" com.sun.javacard.offcardverifier.Verifier -nobanner $@
}

scriptgen() {
	java -Djc.home="$JC_HOME_SIM" -classpath "$SIM_CP" com.sun.javacard.scriptgen.Main -nobanner $@
}

script=/tmp/script

verifycap -outfile /tmp/hello.cap.digest /files/hello.cap
scriptgen -hashfile /tmp/hello.cap.digest -o /tmp/hello.cap.script /files/hello.cap
cat << EOF > $script
powerup;
output off;

> 0x00 0xA4 0x04 0x00 0x09 0xA0 0x00 0x00 0x00 0x62 0x03 0x01 0x08 0x01 0x7F;

Reviewer: ```suggestion
0x00 0xA4 0x04 0x00 0x09 0xA0 0x00 0x00 0x00 0x62 0x03 0x01 0x08 0x01 0x7F; # 2
```
Is this the correct location?

0x00 0xA4 0x04 0x00 0x06 0xAA 0xBB 0xCC 0xDD 0xEE 0xAA 0x7F;
0x88 0x88 0x00 0x00 0x30 $FLAG 0x7f;
0x00 0xA4 0x04 0x00 0x09 0xA0 0x00 0x00 0x00 0x62 0x03 0x01 0x08 0x01 0x7F;
EOF

TMPDIR=/jctmp
mkdir $TMPDIR
cd /upload
for capfile in *.cap; do
    [ -f "$capfile" ] || continue
	verifycap -outfile "$TMPDIR/$capfile.digest" /upload/*.exp "$capfile" || { echo "verify failed"; exit; }
	scriptgen -hashfile "$TMPDIR/$capfile.digest" -o "$TMPDIR/$capfile.script" "$capfile" || { echo "scriptgen failed"; exit; }
	cat "$TMPDIR/$capfile.script" >> /tmp/script
done
echo "All verification finished"

cat << EOF >> $script
0x80 0xB8 0x00 0x00 0x08 0x06 0xAA 0xBB 0xCC 0xDD 0xEE 0xFF 0x00 0x7F;
0x00 0xA4 0x04 0x00 0x06 0xAA 0xBB 0xCC 0xDD 0xEE 0xFF 0x7F;
output on;
0x88 0x66 0x00 0x00 0x00 0x7f;
EOF

wine 'C:\Program Files\Oracle\Java Card Development Kit Simulator 3.1.0\bin\cref_t1.exe' -nobanner -nomeminfo &
sleep 5
java -Djc.home="$JC_HOME_SIM" -classpath "$SIM_CP" com.sun.javacard.apdutool.Main -nobanner -noatr $script
```

The APDU commands in this script do the following:
1. disable output
2. switch to installer applet
3. upload hello.cap
4. switch to hello applet

> 5. insert flag into eeprom

Reviewer: ```suggestion
5. insert flag into EEPROM
```

        this.register();
    }

    public void process(APDU apdu) {
        byte[] buffer = apdu.getBuffer();
        if (buffer[0] == 0 && buffer[1] == -92) {
            return;
        }
        if (buffer[0] == -120 && buffer[1] == -120) {
            if (buffer[4] != 48) {
                ISOException.throwIt((short)26368);
            }
            if (this.isInit) {
                ISOException.throwIt((short)27014);
            }
            JCSystem.beginTransaction();
            Util.arrayCopy(buffer, (short)5, this.secret, (short)0, (short)48);
            this.isInit = true;
            JCSystem.commitTransaction();
            return;
        }
        if (buffer[0] == -120 && buffer[1] == 102) {
            if (buffer[4] != 48) {
                ISOException.throwIt((short)26368);
            }
            if (!this.isInit) {
                ISOException.throwIt((short)27014);
            }
            JCSystem.beginTransaction();
            Util.arrayCopy(buffer, (short)5, this.secret, (short)0, (short)48);
            JCSystem.commitTransaction();
            return;
        }
    }
}

```

## EEPROM Object Layout
![](/imgs/rwctf23-happycard-imhex.png)

> We've taken the liberty to create a small imHex-Pattern to highlight a few of the important bits of an eeprom dump we obtained during analysing of the challenge.

Reviewer: ```suggestion
We've taken the liberty to create a small imHex-Pattern to highlight a few of the important bits of an EEPROM dump we obtained during analysing of the challenge.
```

> We've taken the liberty to create a small imHex-Pattern to highlight a few of the important bits of an EEPROM dump we obtained during analysing of the challenge.

Author: ```suggestion
We've highlighted a few of the important bits of an EEPROM dump we obtained during analysing of the challenge using the ImHex Hex-Editor.
```

The first highlighted block is a `byte[]` filled with `A`s,  directly followed by the definition of our ExploitApplet.

The second highlighted block is a custom Object we named `Foo`, containing a field `short len = 0x1337`.

> And the final block is the FlagArray we want to read, which isn't completely straight forward, as the java card simulator implements a few features to isolate our ExploitApplet from the Safe.

Reviewer: ```suggestion
And the final block is the flag array we want to read, which isn't completely straight forward, as the java card simulator implements a few features to isolate our ExploitApplet from the Safe.
```

> ### Object references

Reviewer: ```suggestion
### Object References
```

To ensure that applets can not blindly access methods on objects from other applets, cref does not directly store references as pointers.
Instead, pointers are represented by *handles*.
A handle is a `short` value representing the object, similar to a file descriptor in Unix.
When an object handle is looked up, the security context of the caller is compared to the security context of the object behind the handle.
If the caller shouldn't be allowed to perform the access, an exception is thrown.

### Anatomy of an Array
Let's take a good look at our self-defined array. It consists of the following fields:

```c
struct Array { // header length: 8 byte
    u16 objectTag = 0x8000;    // this object represents a byte array
    u8 securityContext = 0x22;
    u16 arrayType = 0x0;       // this is a byte array?
    u8 arrayMagic = 0x1C;      // this is an array
    u16 length = 0x30;
    // end of header, followed by data
    u8 data[length];
};
```

> Note that the FlagArray has a different `securityContext`.

Reviewer: ```suggestion
Note that the flag array has a different `securityContext`.
```

At runtime though, we suddenly turn our input `short` into a `byte` array.
However, things aren't quite so simple.
To export and import libraries, javacard files generate so called *export files* (`.exp`). 
These files contain the signature of the methods you export and are checked at two points:
- when generating or loading the library
- when importing a function from the library

No matter what we do, one of these checks now always fails:
Either the import fails as we use a method taking a `short` but the method in the export file takes a `byte[]` or the export file is not congruent with our library code.
Both of these result in verification errors.

Enter the [*PhiAttack*](https://cardis2021.its.uni-luebeck.de/papers/CARDIS2021_Dubreuil.pdf) by Jean Dubreuil and Guillaume Bouffard.

### The PhiAttack

To circumvent the verification, we simultaneously need to produce a congruent export file and also import the methods as they were exported.
This is not possible with two cap files, but we aren't constrained to just two.

By creating a layout similar to the following (from the paper linked above):

![](/imgs/rwctf23-happycard-phiattack.png)

> we can trick the verifier into accepting all cap files, while *still* having a method performing a type confusion.

Reviewer: ```suggestion
we can trick the verifier into accepting all CAP files, while *still* having a method performing a type confusion.
```

### More Confusion

Rather annoyingly, the simulator also very specifically stores the array's type in each array object and actually checks it on every read.
More specifically, the
- `aaload` opcode checks that you read from a reference array
- `saload` opcode checks that you read from a short array
- `iaload` opcode checks that you read from an int array
- `baload` opcode checks that you are not reading from any of the above. <strong style="font-weight: bold;">wat</strong>.

Therefore, we are free to read bytes from *anything*, provided it is not a `short`, `int` or `Object` array.

Very conveniently, the `Object` header is 2 byte shorter than the `array` header. This allows us to freely control the `length` field, if we manage to interpret an `Object` as an array :)

> ## The code

Reviewer: ```suggestion
## The Code
```
